src/main/java/ch/njol/skript/lang/parser/ParsingStack.java

```diff
@@ -0,0 +1,222 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.lang.parser;
+
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SyntaxElement;
+import ch.njol.skript.lang.SyntaxElementInfo;
+import ch.njol.util.Kleenean;
+
+import java.io.PrintStream;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+/**
+ * A stack that keeps track of what Skript is currently parsing.
+ * <p>
+ * When accessing the stack from within {@link SyntaxElement#init(Expression[], int, Kleenean, SkriptParser.ParseResult)},
+ * the stack element corresponding to that {@link SyntaxElement} is <b>not</b>
+ * on the parsing stack.
+ */
+public class ParsingStack implements Iterable<ParsingStack.Element> {
+
+	private final LinkedList<Element> stack;
+
+	/**
+	 * Creates an empty parsing stack.
+	 */
+	public ParsingStack() {
+		this.stack = new LinkedList<>();
+	}
+
+	/**
+	 * Creates a parsing stack containing all elements
+	 * of another given parsing stack.
+	 */
+	public ParsingStack(ParsingStack parsingStack) {
+		this.stack = new LinkedList<>(parsingStack.stack);
+	}
+
+	/**
+	 * Removes and returns the top element of this stack.
+	 *
+	 * @throws IllegalStateException if the stack is empty.
+	 */
+	public Element pop() throws IllegalStateException {
+		if (stack.isEmpty()) {
+			throw new IllegalStateException("Stack is empty");
+		}
+
+		return stack.pop();
+	}
+
+	/**
+	 * Returns the element at the given index in the stack,
+	 * starting with the top element at index 0.
+	 *
+	 * @throws IndexOutOfBoundsException if the given index does not point to
+	 * 										an element in the stack.
+	 */
+	public Element peek(int index) throws IndexOutOfBoundsException {
+		if (index < 0 || index >= size()) {
+			throw new IndexOutOfBoundsException("Index: " + index);
+		}
+
+		return stack.get(index);
+	}
+
+	/**
+	 * Returns the top element of the stack.
+	 * Equivalent to {@code peek(0)}.
+	 *
+	 * @throws IllegalStateException if the stack is empty.
+	 */
+	public Element peek() throws IllegalStateException {
+		if (stack.isEmpty()) {
+			throw new IllegalStateException("Stack is empty");
+		}
+
+		return stack.peek();
```

Comment:
```suggestion
		if (stack.isEmpty())
			throw new IllegalStateException("Stack is empty");
		return stack.peek();
```
I think this is preferrable